packages/agreement-process/src/model/domain/apiConverter.ts

```diff
@@ -97,3 +108,70 @@ export const apiAgreementDocumentToAgreementDocument = (
   path: input.path,
   createdAt: new Date(),
 });
+
+function fromApiTenantVerifier(verifier: ApiTenantVerifier): TenantVerifier {
+  return {
+    id: verifier.id,
+    verificationDate: new Date(verifier.verificationDate),
+    expirationDate: verifier.expirationDate
+      ? new Date(verifier.expirationDate)
+      : undefined,
+    extensionDate: verifier.extensionDate
+      ? new Date(verifier.extensionDate)
+      : undefined,
+  };
+}
+
+function fromApiTenantRevoker(revoker: ApiTenantRevoker): TenantRevoker {
+  return {
+    id: revoker.id,
+    verificationDate: new Date(revoker.verificationDate),
+    expirationDate: revoker.expirationDate
+      ? new Date(revoker.expirationDate)
+      : undefined,
+    extensionDate: revoker.extensionDate
+      ? new Date(revoker.extensionDate)
+      : undefined,
+    revocationDate: new Date(revoker.revocationDate),
+  };
+}
+
+export const fromApiTenantAttribute = (
+  input: ApiTenantAttribute
+): TenantAttribute => {
+  if (input.certified && !input.verified && !input.declared) {
+    return {
+      type: tenantAttributeType.CERTIFIED,
+      id: unsafeBrandId<AttributeId>(input.certified.id),
+      assignmentTimestamp: new Date(input.certified.assignmentTimestamp),
+      revocationTimestamp: input.certified.revocationTimestamp
+        ? new Date(input.certified.revocationTimestamp)
+        : undefined,
+    };
+  } else if (input.verified && !input.certified && !input.declared) {
+    return {
+      type: tenantAttributeType.VERIFIED,
+      id: unsafeBrandId<AttributeId>(input.verified.id),
+      assignmentTimestamp: new Date(input.verified.assignmentTimestamp),
+      verifiedBy: input.verified.verifiedBy.map(fromApiTenantVerifier),
+      revokedBy: input.verified.revokedBy.map(fromApiTenantRevoker),
+    };
+  } else if (input.declared && !input.certified && !input.verified) {
+    return {
+      type: tenantAttributeType.DECLARED,
+      id: unsafeBrandId<AttributeId>(input.declared.id),
+      assignmentTimestamp: new Date(input.declared.assignmentTimestamp),
+    };
+  } else {
+    throw badRequestError(
+      `Invalid tenant attribute in API request: ${JSON.stringify(input)}`
+    );
+  }
```

Comment:
not sure if this could be more readable using `ts-pattern`

Reply:
It could! I tried to do it but didn't manage to make it work 🥲 I had to implement it as it is because, after a lot of debugging, I found that with `ts-pattern` this same logic was not working: https://github.com/pagopa/interop-be-monorepo/pull/480/commits/f9ab4e54d2f1584b432f6e49d6b7c8d6d3359a47 (I tried both with `P.nonNullable` and `P.not(P.nullish)`. 

Maybe I'll investigate a little bit more and try to write a test for this converter, it worries me because it means that either `ts-pattern` is buggy or there's something I'm missing

Reply:
I was able to fix it: https://github.com/pagopa/interop-be-monorepo/pull/480/commits/34332c07f4044cd7de85e1551be88f7fd226de6e

I had to use `P.optional(P.nullish)` because `P.nullish` does not match cases where the key is not set 🥲